src/clients/github.js

```diff
@@ -0,0 +1,280 @@
+  );
+
+  await performWithRetryNetworkErrors(
+    () => createJsFile(github, project, userName, REPO_NAME),
+  );
+
+  await performWithRetryNetworkErrors(
+    () => createPreviewFile(github, project, userName, REPO_NAME),
+  );
+
+  await performWithRetryNetworkErrors(
+    () => updateRepoDescription(github, userName, REPO_NAME),
+  );
+
+  return data;
+}
+
+export async function createGistFromProject(project, user) {
+  const github = clientForUser(user);
+
+  const gist = buildGistFromProject(project);
+  if (isEmpty(gist.files)) {
+    return Promise.reject(new EmptyGistError());
+  }
+
+  const response =
+    await performWithRetryNetworkErrors(() => github.getGist().create(gist));
+
+  const gistData = response.data;
+  if (canUpdateGist(user)) {
+    return updateGistWithImportUrl(github, gistData);
+  }
+  return gistData;
+}
+
+export async function loadGistFromId(gistId, user) {
+  const github = clientForUser(user);
+  const gist = github.getGist(gistId);
+  const response =
+    await performWithRetryNetworkErrors(() => gist.read(), {retries: 3});
+  return response.data;
+}
+
+function buildGistFromProject(project) {
+  const files = {};
+  if (trim(project.sources.html)) {
+    files['index.html'] = {
+      content: project.sources.html,
+      language: 'HTML',
+    };
+  }
+  if (trim(project.sources.css)) {
+    files['styles.css'] = {
+      content: project.sources.css,
+      language: 'CSS',
+    };
+  }
+  if (trim(project.sources.javascript)) {
+    files['script.js'] = {
+      content: project.sources.javascript,
+      language: 'JavaScript',
+    };
+  }
+  if (project.enabledLibraries.length || project.hiddenUIComponents.length) {
+    files['popcode.json'] = {
+      content: createPopcodeJson(project),
+      language: 'JSON',
+    };
+  }
+
+  return {
+    description: 'Exported from Popcode.',
+    'public': true,
+    files,
+  };
+}
+
+async function updateGistWithImportUrl(github, gistData) {
+  const gist = github.getGist(gistData.id);
+  const uri = document.createElement('a');
+  uri.setAttribute('href', '/');
+  uri.search = `gist=${gistData.id}`;
+
+  const description = `${gistData.description} Click to import: ${uri.href}`;
+  const response =
+    await performWithRetryNetworkErrors(() => gist.update({description}));
+  return response.data;
+}
+
+function createPopcodeJson(project) {
+  const json = {};
+  if (project.enabledLibraries.length) {
+    json.enabledLibraries = project.enabledLibraries;
+  }
+  if (project.hiddenUIComponents.length) {
+    json.hiddenUIComponents = project.hiddenUIComponents;
+  }
+  return JSON.stringify(json);
+}
+
+function performWithRetryNetworkErrors(perform, options = {}) {
+  return performWithRetries(
+    perform,
+    errorMessage => errorMessage === NETWORK_ERROR,
+    options,
+  );
+}
+
+async function createSourceFileInRepo(
+  github,
+  userName,
+  repoName,
+  branchName,
+  fileName,
+  source,
+) {
+  return github.getRepo(userName, repoName).
```

Comment:
So this `return` here works fine and does the right thing, because if an `async` function body returns a Promise object explicitly, then that becomes the await-able Promise and all is good.

HOWEVER, I think from a semantic standpoint, we should `await` rather than `return`, because the resolved *value* of the function (which is to say, the resolved value of the promise returned by `writeFile`) is not meaningful. `await` clearly communicates that “this asynchronous operation needs to complete for me to be considered done”, rather than `return` which implies that there is a result value here that someone might care about. It also helps make the `createHtmlFile` etc. functions a bit more concise, as I will mention below…